src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+    gsl_integrand_.function = &detail::integrand<IntegrandType>;
+    gsl_integrand_.params = &integrand;
+    return &gsl_integrand_;
+  }
+
+  size_t max_intervals_;
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+};
+
+void check_status_code(int status_code);
+
+void disable_gsl_error_handling() noexcept;
+
+}  // namespace detail
+
+/// Defines what algorithm is used and should be chosen according
+/// to the problem.
+enum class GslIntegralType {
+  StandardGaussKronrod,            // gsl_integration_qag()
+  IntegrableSingularitiesPresent,  // gsl_integration_qags()
+  IntegrableSingularitiesKnown,    // gsl_integration_qagp()
+  InfiniteInterval,                // gsl_integration_qagi()
+  UpperBoundaryInfinite,           // gsl_integration_qagiu()
+  LowerBoundaryInfinite            // gsl_integration_qagil()
+};
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A wrapper around the GSL adaptive integration procedures
+ *
+ * All templates take upper bounds to the absolute and relative error;
+ * `tolerance_abs` and `tolerance_rel(default = 0.0)`, respectively.
+ * For details on the algorithm see the GSL documentation on `gsl_integration`.
+ *
+ * Here is an example how to use this class. For the function:
+ * \snippet Test_GslQuadAdaptive.cpp integrated_function
+ * the integration should look like:
+ * \snippet Test_GslQuadAdaptive.cpp integration_example
+ */
+template <GslIntegralType TheIntegralType>
+class GslQuadAdaptive;
+
+/*!
+ * The algorithm for "StandardGaussKronrod" uses the QAG algorithm to employ an
+ * adaptive Gauss-Kronrod n-points integration rule. Its function takes a
+ * `lower_boundary` and `upper_boundary` to the integration region, an upper
+ * bound for the absolute error `tolerance_abs` and a `key`. The latter is an
+ * index to the array [15, 21, 31, 41, 51, 61], where each element denotes how
+ * many function evaluations take place in each subinterval. The GSL
+ * documentation mentions that a higher-order rule serves better for smooth
+ * functions, whereas a lower-order rule saves time for functions with local
+ * difficulties, such as discontinuities.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::StandardGaussKronrod>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double upper_boundary, const double tolerance_abs,
+                    const int key, const double tolerance_rel = 0.) const {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qag(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_, key,
+        this->workspace_.get(), &result, &error);
```

Comment:
you don't do anything with `error`. Why not?

Reply:
The first approach was to return the result and the error as a pair. But while discussing with @wthrowe we came to the conclusion that it never ever need to be looked at, since the function throws an error when the given `tolerance_abs` is not reached.

Reply:
Sounds good, thanks!